gradle-conjure/src/main/java/com/palantir/gradle/conjure/ConjureRunnerResource.java

```diff
@@ -0,0 +1,232 @@
+
+            return Optional.of(mainClass.getMethod("main", String[].class));
+        } catch (ReflectiveOperationException e) {
+            log.warn("Failed too get main method {}", mainClassName, e);
+            return Optional.empty();
+        }
+    }
+
+    private static final class ExternalProcessConjureRunner implements ConjureRunner {
+
+        private final File executable;
+
+        ExternalProcessConjureRunner(File executable) {
+            this.executable = executable;
+        }
+
+        @Override
+        public boolean isValid() {
+            // always valid
+            return true;
+        }
+
+        @Override
+        public void close() {
+            // nop
+        }
+
+        @Override
+        public void invoke(Project project, String failedTo, List<String> unloggedArgs, List<String> loggedArgs) {
+            ByteArrayOutputStream output = new ByteArrayOutputStream();
+
+            List<String> combinedArgs = ImmutableList.<String>builder()
+                    .add(executable.getAbsolutePath())
+                    .addAll(unloggedArgs)
+                    .addAll(loggedArgs)
+                    .build();
+
+            ExecResult execResult = project.exec(execSpec -> {
+                project.getLogger().info("Running with args: {}", loggedArgs);
+                execSpec.commandLine(combinedArgs);
+                execSpec.setIgnoreExitValue(true);
+                execSpec.setStandardOutput(output);
+                execSpec.setErrorOutput(output);
+            });
+
+            int exitValue = execResult.getExitValue();
+            log.debug("Executable {} completed with status {} output:\n{}", executable.getName(), exitValue, output);
+
+            if (exitValue != 0) {
+                throw new RuntimeException(String.format(
+                        "Failed to %s. The command '%s' failed with exit code %d. Output:\n%s",
+                        failedTo, combinedArgs, exitValue, output.toString()));
+            }
+        }
+    }
+
+    // We run java things *in-process* to save JVM startup time and reuse JIT optimization (helpful if there are 100
+    // conjure projects)
+    private static final class InProcessConjureRunner implements ConjureRunner {
+
+        private final File executable;
+        private final long executableLastModifiedTime;
+        private final Method mainMethod;
+        private final URLClassLoader classLoader;
+
+        InProcessConjureRunner(File executable, Method mainMethod, URLClassLoader classLoader) {
+            this.executable = executable;
+            this.mainMethod = mainMethod;
+            this.classLoader = classLoader;
+            this.executableLastModifiedTime = executable.lastModified();
+        }
+
+        @Override
+        public boolean isValid() {
+            // If the file has been modified, this data is no longer valid;
+            return executableLastModifiedTime == executable.lastModified();
+        }
```

Comment:
is this isValid stuff _strictly_ necessary? the executable includes the version number in its path right?

Reply:
We can probably remove it, I just wanted to be careful. The most likely case is that something has deleted the location where we extracted our dist. Now that we're using shared services it's less likely.